libqtile/widget/graph.py

```diff
@@ -91,34 +92,43 @@ def graphwidth(self):
     def graphheight(self):
         return self.bar.height - self.margin_y * 2 - self.border_width * 2
 
+    @property
+    def step(self):
+        return max(int(self.graphwidth / float(self.samples)), 1)
```

Comment:
If you set the widget width to, say, 200 and leave samples at 100 then this doesn't work. `graphwidth` will be just less than 200, samples is 100 so step is 1. This means the graph only draws halfway across the widget.

Reply:
You're right. Any suggestions? I don't even use graphs :/

The original problem was that when `cairo` draws multiple operations together(like when we stage all our drawing operations to `RecordingSurface` in `Drawer`) it does some sort of subpixel rendering when 2 vectors touch on the same pixel. The result of that was that `Graph`s suddenly had weird rendering artifacts in the form of blending at the edges of each individual sample.

Reply:
I'll look into if there's some way I can change the code drawing on the surface to handle this better. The only other alternative is to keep track of the rounding errors and add/remove them from a sample at the end to make everything even.

Reply:
I would do it the other way round. Let the user define `step` and calculate the `width` and number of `samples` needed based on that. 

It's the only way I can see that you'll always get the right fit.

Reply:
Re-reading this - my last comment doesn't quite make sense!

Reply:
The options I can see are:
1) Let the user define `width` and `step` and we calculate the number of `samples` needed off the back of that; or 
2) Define `step` and `samples` and calculate the `width`.

Either of those options will result in an integer `step`. The third option, defining `width` and `samples`, will never guarantee this unless there is some manipulation of those values behind the scenes which effectively makes it the same as 1 or 2.

Based on that, option 1 would be my preferred route.

My guess is that `step` will almost always be 1. Removing it as a configurable option would simplify the issue but it would be nicer to leave for users to change if they want.

Reply:
I don't use `Graph` either but I like fixing problems.

Reply:
I'd like to avoid a changelog entry for this if possible... I would prefer if I can fix it without changing the graph configuration

Reply:
of course if it's impossible then 🤷